### PR TITLE
Add `.amdhsa_code_object_version` metadata serializing rocDL modules

### PR DESCRIPTION
**Pull request by umangyadav**
Fixes https://github.com/ROCm/rocMLIR-internal/issues/1205

### CODE REVIEW

File: external/llvm-project/mlir/include/mlir/Target/LLVM/ROCDL/Utils.h [hunk 1]
```diff
@@ -68,17 +68,19 @@ class SerializeGPUModuleBase : public LLVM::ModuleToObject {
   LogicalResult appendStandardLibs(AMDGCNLibraries libs);
 
   /// Loads the bitcode files in `fileList`.
-  virtual std::optional<SmallVector<std::unique_ptr<llvm::Module>>>
+  std::optional<SmallVector<std::unique_ptr<llvm::Module>>>
```
Comment (krzysz00) [inline external/llvm-project/mlir/include/mlir/Target/LLVM/ROCDL/Utils.h]: Put the `virtual` back - we need it downstream

File: external/llvm-project/mlir/lib/Target/LLVM/ROCDL/Target.cpp [hunk 2]
```diff
@@ -271,10 +273,9 @@ void SerializeGPUModuleBase::addControlVariables(
                        isaVersion.Minor + 100 * isaVersion.Stepping +
                            1000 * isaVersion.Major,
                        32);
-    int abi = 500;
-    abiVer.getAsInteger(0, abi);
     addControlVariable("__oclc_ABI_version", abi, 32);
   }
+  addControlVariable(".amdhsa_code_object_version", abi, 32);
```
Comment (krzysz00) [inline external/llvm-project/mlir/lib/Target/LLVM/ROCDL/Target.cpp]: Nope!

We want a module flag

And this is independent of device library loading

Comment (krzysz00) [reply external/llvm-project/mlir/lib/Target/LLVM/ROCDL/Target.cpp]: And this won't run if no device libs are being loaded per the if statement at the top of this function

**Review by krzysz00 — CHANGES_REQUESTED** [review]
You need to set metadata on the relevant module, not just add to the device library handling code.

Also, test updates?

File: external/llvm-project/mlir/include/mlir/Target/LLVM/ROCDL/Utils.h [hunk 3]
```diff
@@ -71,14 +71,16 @@ class SerializeGPUModuleBase : public LLVM::ModuleToObject {
   virtual std::optional<SmallVector<std::unique_ptr<llvm::Module>>>
   loadBitcodeFiles(llvm::Module &module) override;
 
-  /// Adds `oclc` control variables to the LLVM module.
+  /// Adds `oclc` and `code_object_version` control variables to the LLVM
```
Comment (krzysz00) [inline external/llvm-project/mlir/include/mlir/Target/LLVM/ROCDL/Utils.h]: I'd say
```
/// Adds `oclc` control variables to the LLVM module if needed and sets the `amdhsa_code_object_version` flag.
```
or something like that

**Review by krzysz00 — COMMENTED** [review]
Overall seems right mod test coverage per our DMs

File: external/llvm-project/mlir/unittests/Target/LLVM/SerializeROCDLTarget.cpp [hunk 4]
```diff
@@ -114,7 +114,8 @@ TEST_F(MLIRTargetLLVMROCDL, SKIP_WITHOUT_AMDGPU(SerializeROCDLToPTX)) {
   ASSERT_TRUE(!!module);
 
   // Create a ROCDL target.
-  ROCDL::ROCDLTargetAttr target = ROCDL::ROCDLTargetAttr::get(&context);
+  ROCDL::ROCDLTargetAttr target = ROCDL::ROCDLTargetAttr::get(
+      &context, 2, "amdgcn-amd-amdhsa", "gfx900", "", "400");
```
Comment (krzysz00) [inline external/llvm-project/mlir/unittests/Target/LLVM/SerializeROCDLTarget.cpp]: I'd suggest making a copy of the test, one that checks for the current default, and one that checks for 400.

**Review by krzysz00 — APPROVED** [review]
Overall, approved, glad we're getting this handled

File: external/llvm-project/mlir/unittests/Target/LLVM/SerializeROCDLTarget.cpp (hunk 4 above)
Comment (umangyadav) [reply external/llvm-project/mlir/unittests/Target/LLVM/SerializeROCDLTarget.cpp]: Done

**Comment by umangyadav** [comment]
@krzysz00  changed the commit history, let me know if it looks okay

**Comment by krzysz00** [comment]
@umangyadav In trivialities:

- Move the `external/` typo fix into the external-facing commit
- Have the external/ part have a commit message that starts with something like `[mlir][ROCDL] Add code object version metadata when `serializing modules` (feel free to play with the wording) so that we can cherry-pick it right into an upstream PR

**Comment by umangyadav** [comment]
> @umangyadav In trivialities:
> 
> * Move the `external/` typo fix into the external-facing commit
> * Have the external/ part have a commit message that starts with something like `[mlir][ROCDL] Add code object version metadata when `serializing modules` (feel free to play with the wording) so that we can cherry-pick it right into an upstream PR

Done. Thanks

**Pull request merged**